src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/Sources/ManualResetValueTaskSourceCore.cs

```diff
@@ -211,47 +211,68 @@ private void SignalCompletion()
             {
                 if (_executionContext != null)
                 {
-                    ExecutionContext.RunInternal(
-                        _executionContext,
-                        (ref ManualResetValueTaskSourceCore<TResult> s) => s.InvokeContinuation(),
-                        ref this);
+                    // ExecutionContext flow enabled
+                    Thread currentThread = Thread.CurrentThread;
+                    // Restore the ExecutionContext before executing anything.
+                    ExecutionContext? currentContext = ExecutionContext.RestoreInternal(_executionContext, currentThread);
```

Comment:
Why does it need to use RestoreInternal rather than Restore?  If that's necessary for perf, doesn't that suggest the proposed ExecutionContext.Restore is insufficient?

Reply:
Save on cost of accessing the `Thread.CurrentThread` thread static twice; also returns the current context so can avoid `EC.Capture()` (which also uses the thread static `Thread.CurrentThread`)

> doesn't that suggest the proposed ExecutionContext.Restore is insufficient?

The use case I have for it is calling  `EC.Capture()` once outside loop; and then `EC.Restore()` at end of each loop iteration. The per-iteration Restore would access the thread static just as many times as using `EC.Run` would so it wouldn't change that dynamic.

Exposing the api that takes the an additional currentThread parameter would make it unsafe, as it's not necessarily the same thread after an `await`, so we can't capture it then reuse it, whereas here it's always the same thread because its not async, so we can take that shortcut.

Reply:
I understand, but if EC.Restore(EC) is insufficient for MRVTSC, why do we believe it's sufficient for other cases that would cause us to make it public?  MRVTSC is the only current consumer of the alternative callback-based API, and so for me, this is the test case as to whether exposing an EC.Restore(EC) is actually viable.  If it doesn't work here but we expose the method anyway, I fully expect to hear from you again about not being able to use EC.Restore(EC) elsewhere successfully because it adds XYZ overheads and needing a different solution.

Reply:
Just using the simple case now :)